frame/support/src/traits/tokens/fungibles/regular.rs

```diff
@@ -584,9 +584,16 @@ pub trait Balanced<AccountId>: Inspect<AccountId> + Unbalanced<AccountId> {
 	fn done_withdraw(_asset: Self::AssetId, _who: &AccountId, _amount: Self::Balance) {}
 }
 
-/// Use an on-chain exchange to convert the asset to the equivalent in native tokens.
-pub trait SwapForNative<Origin, AccountId, Balance, AssetBalance, AssetId> {
-	// If successful returns the amount in native tokens.
+/// Trait for providing methods to swap between the chain's native token and other asset classes.
+pub trait SwapNative<Origin, AccountId, Balance, AssetBalance, AssetId> {
+	/// Take an `asset_id` and swap some amount for `amount_out` of the chain's native asset. If an
```

Comment:
Do you actually need all of this to be generics, instead of associated types? You can probably save a bunch of boilerplate if you revise. 

From what I can seem we want the asset-related stuff to be generic, but everything else to be associated types, because a chain might have multiple asset classes, but almost always has just one native currency. So something like

```
impl SawpNative<AssetType1> for PalletA {...}
impl SawpNative<AssetType2> for PalletB {...}
```

is reasonable but I don't think we ever want to customize the origin per implementation.

Reply:
I would invite @gilescope to answer your question :) It was his idea to create this generic